refactor(common): remove non-null assertions from directives

JeanMeche opened the pull request:
## PR Checklist
Per request from @pkozlowski-opensource, I'm splitting PR #48241 into 2. 

See individual commits. 

These are fixes targeted to resolve https://github.com/angular/angular/issues/24571

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [x] No

JeanMeche committed:
refactor(common): remove non-null assertions

As part of angular#24571, removing all non-null assertions from common/directives

JeanMeche committed:
refactor(common): remove non-null assertions

As part of angular#24571, removing all non-null assertions from common/directives

JeanMeche committed:
refactor(common): code cleaning

* removing guard as console.warn is now widely supported
* Couldn't remove non-null assertion, waiting for TS support of getters with different types (microsoft/typescript#43662)

atscott commented on packages/common/src/directives/ng_plural.ts:
```diff
@@ -71,8 +68,10 @@ export class NgPlural {
     const cases = Object.keys(this._caseViews);
-    const key = getPluralCategory(this._switchValue, cases, this._localization);
-    this._activateView(this._caseViews[key]);
+    if (this._switchValue !== undefined) {
```
Would `_switchValue` be better with a default value? It's actually kind of correct with the `!` right now. The directive won't match unless this is defined. It's a little unfortunate that we have to _add_ code to account for the possibility of `undefined` when it really can't ever be in production.

JeanMeche replied:
You're right, we can keep the type safety by passing the value to `_updateView`

atscott commented on packages/common/src/directives/ng_switch.ts:
```diff
@@ -106,8 +106,7 @@ export class SwitchView {
 export class NgSwitch {
-  // TODO(issue/24571): remove '!'.
-  private _defaultViews!: SwitchView[];
+  private _defaultViews?: SwitchView[];
```
Should this have a default of empty array instead of undefined?

JeanMeche replied:
I'd rather see undefined or null as a default value.  It's created when needed by `_addDefault` and checked by `_updateDefaultCases`. 

I'm always to other point of views though.

atscott replied:
Having it be nullable rather than empty list provides no additional meaning. I feel quite strongly that this should never be `null` or `undefined`. You can also update the rest of the code to remove the truthiness checks of this value.

JeanMeche replied:
I used empty array as default value. It seems presubmit is failing now with the changes. Any heads-up ?

atscott reviewed (approved):
👍 LGTM

atscott commented on packages/common/src/directives/ng_plural.ts:
```diff
@@ -49,29 +48,27 @@ import {SwitchView} from './ng_switch';
-  private _switchValue!: number;
-  // TODO(issue/24571): remove '!'.
-  private _activeView!: SwitchView;
+  private _switchValue?: number;
```
It doesn't look like this is used anywhere now that the value is passed to `_updateView`. Let's remove it.

JeanMeche replied:
done 👍

atscott commented on packages/common/src/directives/ng_switch.ts:
```diff
@@ -149,10 +145,9 @@ export class NgSwitch {
 
   private _updateDefaultCases(useDefault: boolean) {
-    if (this._defaultViews && useDefault !== this._defaultUsed) {
+    if (useDefault !== this._defaultUsed) {
```
```suggestion
    if (this._defaultViews.length > 0 && useDefault !== this._defaultUsed) {
```
I think you need to have the `length` check here to be equivalent with how it was before.

JeanMeche replied:
Done. I also cleaned up the commits to match 1 file changed/per commit.

atscott reviewed (approved):
@JeanMeche looks like the lint test is failing. The code changes otherwise look good and the internal test failures are unrelated

JeanMeche committed:
refactor(common): remove non-null assertions

As part of angular#24571, removing all non-null assertions from common/test/directives

alxhub commented:
This PR was merged into the repository by commit d887d69f241f67dd77c8533f7eed08d6397f98a8.
